### ISSUE TITLE
New contract interface and  EIPs should be required to include a privacy considerations section

### ISSUE THREAD

**Issue by kdenhartog** [ethereum/EIPs#5682]
### Proposed Change

This one was a bit obvious to me and it's only once I started reading through EIP-1 that I realized it's not required. Pretty simple to remediate. Let's just add a requirement that new EIPs being created and any EIPs entering the final call stage are required to include a Privacy considerations section. A reference to how to help developers define this section can be linked to here: https://datatracker.ietf.org/doc/rfc6973/

**Comment by Pandapip1** [comment]
I think this should be an (optional?) subsection of the security considerations section. Privacy is a security consideration, and all non-obvious privacy considerations should be listed there.

There are plenty of EIPs where privacy considerations are not an issue. On the blockchain, everything is public by default, and everything should be assumed to be public except when commonly known to be private (e.g. private keys) or explicitly stated.

**Comment by kdenhartog** [comment]
Privacy and security are often separate issues to be considered. This is why IETF cites the necessity for both of them. Even in the case where no known privacy or security issue is known a simple statement can be stated as such. For example, a document outlining how to define privacy considerations still defines a security considerations section even though it's defining how to write a document. https://www.rfc-editor.org/rfc/rfc6973.html#section-9

I don't think it's unreasonable to require an author to include these sections. Even if it's as simple as stating that nothing needs to be considered.

**Comment by Pandapip1** [comment]
> I don't think it's unreasonable to require an author to include these sections. Even if it's as simple as stating that nothing needs to be considered.

The *vast* majority of EIPs don't need this section. As previously mentioned, everything is public by default. Could you point to a specific EIP in which this section would be useful?

**Comment by xinbenlv** [comment]
@kdenhartog thank you for the proposal, can you give us a bit of examples based on some EIP what their privacy section will look like?

**Comment by kdenhartog** [comment]
I tend to agree that core category EIPs don't seem to have any need for these considerations. I've not found one so far as I've briefly skimmed them and come up with some things that could be discussed in a few EIPs.

Starting from the beginning. Here's just a few things I'd be thinking about to enhance the privacy of these different EIPs if I was considering implementing them. There's plenty more and I can keep going if you think it would be useful to show how this would be useful.

EIP-20: Considerations around the usage of too large of a decimal value allowing too much precision which can be used to correlate transfers. Considerations around linkability of transfers between users and well known addresses such as exchanges.

EIP-86 (and various other account abstraction works): considerations how various types of contract logic such as multisignatures and delegation links two persistant identifiers (e.g. ethereum accounts)

EIP-107: Fingerprinting concerns to identify the user as a web3 user.

EIP-137: linking of persistant identifiers. Fingerprinting concerns. Usage of a new persistant identifier. Basically all the did-core privacy considerations apply here too: https://www.w3.org/TR/did-core/#privacy-considerations

EIP-162: Same considerations apply as 137

EIP-173: correlating activity done as a contract owner with other on chain activity. E.g. should a personal be used when deploying the contract separate from an account used for interacting with other contracts?

EIP-181: probably lots of similar considerations as 137 and 162 here.

EIP-191: Displaying warnings to users about checking to make sure no PII is included in the transactions

EIP-600: Considerations around the ability for people with access to a public key being able to derive child path keys and link ownership of ETH accounts

EIP-601: Same as 600

EIP-627: Usage of this protocol for a P2P messaging scheme to link two communicating parties by a passive malicious actor.

EIP-634: Lol this one is funny because it's now encouraging storing PII on chain which has generally been regarded as a bad idea due to the immutability property

**Comment by xinbenlv** [comment]
I see. Thank you for a very long list of examples. That's very helpful. I am weakly in favor of adding a privacy consideration section, and make it optional. I do think that the complexity of privacy consideration also lies in the different expectations of web3 vs web2...

**Comment by Pandapip1** [comment]
Okay, I think that's a big enough list to justify it being its own 2nd level section. I still don't think it's large enough to warrant being a required section though.

**Comment by kdenhartog** [comment]
I'd be in favor of requiring for Interface/ERC but optional for network and core. The purpose really is to force the authors to consider the adversarial aspects of privacy invasions and to spur the discussion. If there truly isn't any than a quick short sentence would be suffice I think.

**Comment by Pandapip1** [comment]
I still don't think it should be a required section for ERC category EIPs, if only because the expectations of privacy on the blockchain are greatly lowered. For interface EIPs, though, I would be in favor of requiring it since most of them do introduce privacy risks when using web2.

**Comment by kdenhartog** [comment]
Main reason I saw them as useful is to highlight how clients should be interacting with the ERCs. Assuming a user interacted with a single discrete contract the risk is quite minimal. However, this is an uncommon pattern in this space so making recommendations to wallet developers is that are interacting with these standard contracts will assist with how they go about implementing support for the contracts. Once the data is public I agree there's little that can be done. Including this section as a required option can open up a discussion that justifies a modification to an interface to reduce the amount of information that gets published.

For example of this see the recent privacy considerations section I've been working on with the author of account bound tokens in #5686

It significantly changes how users would interact with a contract to expand the definition beyond an interface to also include expected behavior such as including off chain URLs rather than the data directly on chain.

**Comment by SamWilsn** [comment]
None of the active editors are lawyers, so I don't think we're qualified to review privacy concerns.

---

I don't think any of the listed EIPs have enough privacy concerns to justify more than a sentence in the Security Considerations section. EIPs that specifically deal with PII should certainly have a privacy subsection, but generally speaking EIPs deal with interfaces, algorithms, and pseudonymous accounts; not personal information.

**Comment by kdenhartog** [comment]
> None of the active editors are lawyers, so I don't think we're qualified to review privacy concerns

Privacy concerns don't have to have a legal element to them? In fact, that's one of the things being discussed in #5686 is that GDPR isn't a good basis for why a technical consideration is considered. 

This is how W3C handles privacy considerations sections and horizontal reviews of them: 

https://www.w3.org/Guide/documentreview/#how_to_get_horizontal_review

**Comment by Pandapip1** [comment]
> https://www.w3.org/Guide/documentreview/#how_to_get_horizontal_review

Side note - the EIP process really needs a horizontal review process like that.

**Comment by Joumax11** [comment]
Tanks for all helps

**Comment by Pandapip1** [comment]
This is definitely an important issue and should be kept open.

**Comment by kdenhartog** [comment]
I think the general consensus for this has landed on keep it optional and I'm good with that (Personally, I think requiring it would be more beneficial, but I'm willing to compromise). The part that's not clear is whether this should added as a new top level section or if this should be made a subsection of the security considerations section. It seems that making it a subsection of security implies that privacy is a subset of security considerations when often times they are considered separate. For example, correlation of an EoA account across different dApps isn't a security issue because it's a public key that has no effect on the outcome of the security model. However, it does allow for the user to be tracked across dApps which does impact privacy still. Therefore, it makes sense that if we want this section to be added to the template it should be considered a separate top level section that is optional to include.

**Comment by Pandapip1** [comment]
I personally would prefer a 2nd-level "Considerations" with 3rd-level "Security", "Privacy", and "Other".

**Comment by xinbenlv** [comment]
+1 

Agree with @kdenhartog that we want security and privacy considerations in the same level. 

Agree with @Pandapip1 that it's better to have a top level "Considerations" with security and privacy as second level section.

**Comment by kdenhartog** [comment]
I'm +1 to that, that seems like something that could achieve consensus and then lead naturally to a path overtime as expertise is built up. I'm willing to update the template and Walidator bot if the other editors are in alignment with this as well.

My idea for how this might play out is that for we could make all these sections optional. Then overtime if we build up proper expertise for security/privacy (accessibility is another horizontal review process that could be useful later on) "review boards" that can help review and provide feedback to these sections. At that point it may make sense to turn these sections from optional to required, but until we have sufficient number of people with expertise willing to help I think this is the best middle ground for now.

**Comment by Pandapip1** [comment]
Waiting on consensus

**Comment by andrey-savov** [comment]
> I think this should be an (optional?) subsection of the security considerations section. Privacy is a security consideration, and all non-obvious privacy considerations should be listed there.

@Pandapip1 Privacy and security overlap but neither is strict subset of the other.  Lumping both types of considerations into one section would be a mistake.  In general privacy is about a person while security is about data.

**Comment by Pandapip1** [comment]
> @Pandapip1 Privacy and security overlap but neither is strict subset of the other. Lumping both types of considerations into one section would be a mistake. In general privacy is about a person while security is about data.

There would be subsections. So:

---

## Considerations

### Security

(Security considerations here)

### Privacy

(Privacy considerations here)

### Other

(Other considerations here)

**Comment by Pandapip1** [comment]
Waiting on consensus.

**Comment by Pandapip1** [comment]
Dismissing stale bot

**Comment by kdenhartog** [comment]
re-pinging @axic @gcolvin @lightclient @SamWilsn to see if they've got an opinion on this. If they don't weigh in within the next 2 weeks I think we should assume that they've seen this and chosen to abstain from the consensus process. Wdyt about that @Pandapip1?

**Comment by Pandapip1** [comment]
I have agreed to a version of this: https://github.com/ethereum/EIPs/issues/5682#issuecomment-1364578967

**Comment by kdenhartog** [comment]
I'm in alignment with you on that as well. I think it's just the other 4 editors that we're struggling to get opinions from.

**Comment by kdenhartog** [comment]
Oh ok, I didn't realize governance rules are currently set around this. Or are you saying that's just how we should do it at this point because we don't have any rules set around how to make changes around this?

**Comment by Pandapip1** [comment]
> Oh ok, I didn't realize governance rules are currently set around this. Or are you saying that's just how we should do it at this point because we don't have any rules set around how to make changes around this?

Technically the quorum is two, but I feel like this is a significant enough change to warrant three editors.

@SamWilsn is the next most active editor. How does https://github.com/ethereum/EIPs/issues/5682#issuecomment-1364578967 look?

**Comment by Pandapip1** [comment]
Dismissing stale bot

**Comment by Pandapip1** [comment]
Dismissing stale bot

**Comment by kdenhartog** [comment]
Thanks for summarizing this @xinbenlv to make it easier for other editors to consider.

I'd be +1 to any of these options

**Comment by SamWilsn** [comment]
I'm not opposed to including privacy concerns. I'm opposed to adding a whole section or subsection for it.

Until someone convinces me that there exists a privacy concern that isn't also a security concern, privacy concerns should go in Security Considerations.

**Comment by kdenhartog** [comment]
> I'm not opposed to including privacy concerns. I'm opposed to adding a whole section or subsection for it.
> 
> Until someone convinces me that there exists a privacy concern that isn't also a security concern, privacy concerns should go in Security Considerations.

Not sure if you saw it, but I listed a few above in this comment https://github.com/ethereum/EIPs/issues/5682#issuecomment-1251624043

Fingerprinting and identifier correlation are the most common class of issues that present no security risks, but introduce privacy risks. So for example, https://fingerprint.com/ is a tool used to deanonymize users on the web without needing to create a persistent identifier like a cookie or getting the user to submit their email address in order to cross site track them. In Brave, we implement specific defenses to prevent fingerprinting described here: https://brave.com/privacy-updates/4-fingerprinting-defenses-2.0/

Additionally, in the case of web3 we have a very bad problem where all identifier de-correlation falls on the user to separate out their tokens into many different accounts. Since this presents additional costs to users and doesn't align with the open world protocol ecosystem, most users end up just reusing the same account across many different DApps in a way that introduces a globally unique, persistent identifier that will be used to cross site track them. This is a privacy risk solely, but presents no security harms. It's akin to saying a user using the same email address across many different websites is a privacy risk, but it doesn't grant every site they use it with access to their email account which would be a separate security risk.

**Comment by Pandapip1** [comment]
Dismissing stale bot

**Comment by chaals** [comment]
Some comments:

I agree with @kdenhartog - security and privacy are fundamentally different. If your psychiatrist doesn't give away your address to random people who decide that your condition makes you exploitable, you're secure. But you have a right IMHO to expect that they also don't mention it at a dinner with people who won't actually attack you.

We don't need to be lawyers - privacy isn't well-defined, but it's something that people do understand. And it is complex.

An optional section on privacy as part of considerations would be a good thing. I happen to think that it should be compulsory, even if in many cases it seems like there is no issue. Because in other cases, there really is, and those issues work against achieving goals like broad adoption.

**Comment by kdenhartog** [comment]
bump

**Comment by chaals** [comment]
There seems to be a governance problem here. I can bump this every week to keep it alive. It isn't clear how we reach a decision that seems legitimate to the community at large. So we're going round in circles chewing up time.

Not a good look. I suspect the answer isn't specific to this issue though.

**Comment by Pandapip1** [comment]
The problem is that the EIP process is intentionally slow. The bot is there to remind us which issues exist and to ensure that the only issues that get discussed are ones where there is at least one person that spends time to bump it.

I think @SamWilsn is the only one against this. In the meantime, please feel free to put privacy considerations in the security considerations.

**Comment by xinbenlv** [comment]
@Pandapip1 thanks for comment. I agree with @kdenhartog and @chaals that it's really important to target reader who are builders of the space to distinguish privacy from security. I've signed up a time to discuss this EIPIP Meeting.

**Comment by kdenhartog** [comment]
> The problem is that the EIP process is intentionally slow.

For me, I'm less concerned about the time, I've worked on specs for years at a time with some issues taking years to reach a solution with lots of disagreement along the way. I know @chaals has as well with his time in W3C and IETF. I'd say it's more so the opacity of how to push an issue towards consensus or find common ground that's hard right now. As @chaals points out this isn't necessarily specific to this issue and it's something that I think many of us have been thinking about for a few months now. We'll keep working on that in parallel to this, but should utilize this issue as an example of why it would be useful for us to improve and to help guide us on how to improve our governance.

> I've signed up a time to discuss this EIPIP Meeting.

Perfect, I'll plan to attend EIPIP 77 to try and see if we can find a resolution to this specific issue as well.

**Comment by xinbenlv** [comment]
https://github.com/ethereum-cat-herders/EIPIP/issues/219#issuecomment-1479752070

**Comment by Pandapip1** [comment]
![](https://imgs.search.brave.com/8j-qMt9gHs5eNxfZdzi_EaV2KdVOGz4yI3UI_w_9ySE/rs:fit:711:225:1/g:ce/aHR0cHM6Ly90c2U0/Lm1tLmJpbmcubmV0/L3RoP2lkPU9JUC5J/andHVGhuTjd3ZUNZ/dTJZRE9GWjNBSGFF/OCZwaWQ9QXBp)

**Comment by Skylerspencer** [comment]
I agree 20000%

**Comment by chaals** [comment]
la-la-la

**Comment by Pandapip1** [comment]
Bump

**Comment by chaals** [comment]
bump. It's still unfortunate that there doesn't seem to be an answer to **how** to get this to a decision, let alone any prospect of a decision.

**Comment by SamWilsn** [comment]
Just popping in to reaffirm that:

 * I am against a mandatory privacy considerations section (at any heading level.)
 * I am against an optional top-level privacy considerations section (it should go in the same section as security considerations.)
 * I support a privacy considerations heading somewhere in the `eip-template.md` by default.

**Comment by chaals** [comment]
Security and privacy have some overlaps, like security and accessibility do. However, they are different, so it makes no sense to me to insist that privacy be considered a subset of security.

That said, it seems unconscionable that we just do nothing and leave the privacy implications of new EIPs to individuals' intentions while we drag out this discussion.

So my heirarchy of options:

- Preferred: A top-level, mandatory section on privacy considerations
- Reasonable: A top-level, section on privacy considerations that is obligatory to the same extent as security considerations
- Reasonable: A section somewhere (anywhere) on privacy considerations that is obligatory to the same extent as security considerations
- Acceptable: A top-level, optional section on privacy considerations
- Barely acceptable: An optional section anywhere in the template on privacy considerations
- Unacceptable: Just continue to prolong this discussion.

**Comment by kdenhartog** [comment]
@SamWilsn that makes sense. Are you in support of this format proposed by @Pandapip1 which I believe would align with:
> I support a privacy considerations heading somewhere in the eip-template.md by default.


That seems to have the best consensus and is a middle ground that I think we can at least all get behind to move this forward. Original proposal made here https://github.com/ethereum/EIPs/issues/5682#issuecomment-1364578967
 
> ## Considerations
> ### Security
> (Security considerations here)
> 
> ### Privacy
> (Privacy considerations here)
> 
> ### Other
> (Other considerations here)

Even with your support I still think we need either @axic @lightclient or @gcolvin to voice support to achieve quorum of at least 3 editors. I can't remember if @xinbenlv was granted full status for a vote like this or if he's still on temporary status at this point.

**Comment by SamWilsn** [comment]
> Are you in support of this format proposed by @Pandapip1

No. Any general "other" considerations section is a hard no from me. We already have enough difficulty getting authors to put the correct content in motivation and rationale. Adding a catch-all section where people can write whatever they want just makes it even more ambiguous.

> Security and privacy have some overlaps, like security and accessibility do. However, they are different, so it makes no sense to me to insist that privacy be considered a subset of security.

Can you give me an example of a privacy consideration that isn't a security consideration?

> Even with your support I still think we need either @axic @lightclient or @gcolvin to voice support to achieve quorum of at least 3 editors. I can't remember if @xinbenlv was granted full status for a vote like this or if he's still on temporary status at this point.

Our governance decisions aren't made by a majority vote. As long as an editor is strongly objecting (like me now), nothing changes.

**Comment by kdenhartog** [comment]
> Can you give me an example of a privacy consideration that isn't a security consideration?

See section 5.2 of RFC 6973 for classes of privacy considerations that are not security considerations.

See this comment for more specific examples applied to EIPs https://github.com/ethereum/EIPs/issues/5682#issuecomment-1251624043

> Our governance decisions aren't made by a majority vote. As long as an editor is strongly objecting (like me now), nothing changes.

Where is this governance process defined? I can't seem to find it and keeping it as tribal knowledge is less than ideal. Also, where is it defined how someone in the community can suggest an improvement to this governance process?

**Comment by SamWilsn** [comment]
> See section 5.2 of RFC 6973 for classes of privacy considerations that are not security considerations.

Correlation "increase[s] the power that those doing the correlating have over individuals as well as correlators' ability to pass judgment, threatening individual autonomy [...]". That sounds like a huge issue for personal security.

Similarly, identification "mak[ing] it easier for individuals to be explicitly controlled by others" sounds like another issue for personal security.

At least with the right framing, all these examples seem like issues of personal security to me

**Comment by SamWilsn** [comment]
> Where is this governance process defined? I can't seem to find it and keeping it as tribal knowledge is less than ideal. Also, where is it defined how someone in the community can suggest an improvement to this governance process?

We haven't needed to formalize our governance process. Up until recently, there were only four editors that participated in governance :rofl:

The easiest way to participate in the governance process are discussions like this one, and to attend EIPIP meetings. Beyond those options, you could:

 * Become an EIP editor, which lets you _also_ veto changes to the process, or
 * Fork the repository to run it however you like.

**Comment by SamWilsn** [comment]
@gcolvin [suggested renaming the section to `Security and Privacy`](https://discord.com/channels/916516304972824576/916713912970412103/1110393036950745088). That works for me.

**Comment by kdenhartog** [comment]
making this unstale. Next steps is still on me to update the template and EIP-1

**Comment by chaals** [comment]
tralala

**Comment by SamWilsn** [comment]
Looks like more editors are objecting to this than just me. For the time being, the section will remain "Security Considerations", with the recommendation that privacy concerns go in that section.

**Comment by Pandapip1** [comment]
I, for one, do not object to the renaming.

**Comment by kdenhartog** [comment]
Yeah I saw on the notes that @lightclient was the one who had some reservations on this. I've not gotten a chance to listen through the call yet, but if they could comment here with their concerns that would be awesome so we can consider what the best next steps are to progress this issue.

**Comment by kdenhartog** [comment]
I'm awaiting what happens with core devs since I think they're less inclined to want the additional burden from this right now. With the ERC devs and the wallet devs I think this will be more pertinent though.

**Comment by chaals** [comment]
> ... core devs since I think they're less inclined to want the additional burden from this ...

I understand they might not want to make their work harder (who does?). But it's a bit disappointing, given the central role they play in the ecosystem, to learn the core devs don't take this issue seriously enough that they already do it as a matter of course :(

**Comment by Pandapip1** [comment]
Will reopen this. A consensus has definitely not been reached.

**Comment by chaals** [comment]
@Pandapip1 thanks for reopening before.

This feels like we're relying on "consensus" by attrition. Which is a pretty broken way to run anything, since the thing it mostly shows is that caring about getting something right isn't worth the effort.

I hope, and trust, that isn't actually the case for Eth.

**Comment by kdenhartog** [comment]
My understanding is that after the split this is likely going to be something that has to be pushed via each WG. So I think there's little contention that this is necessary for the ERC and Wallet devs. Unfortunately, I think there's been a bit of push back on this for core devs. I'd still like to see core devs support this as many of their EIPs could still be improved with privacy considerations as pointed out in https://github.com/ethereum/EIPs/issues/5682#issuecomment-1251624043, but we'll likely need to engage directly with them on this now instead of pushing it via EIPIP if I understand it correctly.

**Comment by Pandapip1** [comment]
Yup. This is again why I'm trying to push for a meta-WG, but I feel like that's getting less and less likely.

**Comment by kdenhartog** [comment]
I'd agree that would be useful, but my understanding is Core wants to wait to do that for now. That makes sense to wait until Wallet and ERCs have active WGs otherwise it just becomes ErC/wallet/editor people telling core how to write their EIPs. I can understand why that doesn't align with their goals at the moment but I think in the future it likely could based on a few discussions I've had with some of the core devs.
